### ISSUE TITLE
Occasional failed `require` in `akka.dispatch.BatchingExecutor` during tests (provoked with HTTP 1.0-RC1)

### ISSUE THREAD

**Issue by RichardBradley** [akka/akka#17356]
I am getting the following exception thrown inside my unit tests about 1 run in 5.
I may well be using "BatchingExecutor" or "MarshallingTestUtils" wrong, but I think that the exception message should explain how if so.

```
Caused by: java.lang.IllegalArgumentException: requirement failed
    at scala.Predef$.require(Predef.scala:207)
    at akka.dispatch.BatchingExecutor$BlockableBatch.blockOn(BatchingExecutor.scala:105)
    at scala.concurrent.Await$.result(package.scala:190)
    at akka.http.scaladsl.testkit.MarshallingTestUtils$class.marshal(MarshallingTestUtils.scala:18)
    at myapp.routing.RoutingServiceSpec.marshal(RoutingServiceSpec.scala:33)
    at myapp.routing.RoutingServiceSpec$$anon$1.process(RoutingServiceSpec.scala:59)
    at myapp.layers.ProcessingLayer.apply(ProcessingLayer.scala:41)
    at myapp.routing.RoutingService$$anonfun$dispatchInner$1.apply(RoutingService.scala:405)
    at myapp.routing.RoutingService$$anonfun$dispatchInner$1.apply(RoutingService.scala:399)
    at akka.http.scaladsl.server.directives.RouteDirectives$$anonfun$complete$1.apply(RouteDirectives.scala:45)
    at akka.http.scaladsl.server.directives.RouteDirectives$$anonfun$complete$1.apply(RouteDirectives.scala:45)
    at akka.http.scaladsl.server.StandardRoute$$anon$1.apply(StandardRoute.scala:19)
    at akka.http.scaladsl.server.StandardRoute$$anon$1.apply(StandardRoute.scala:19)
    at akka.http.scaladsl.server.directives.BasicDirectives$$anonfun$mapRouteResult$1$$anonfun$apply$3.apply(BasicDirectives.scala:32)
    at akka.http.scaladsl.server.directives.BasicDirectives$$anonfun$mapRouteResult$1$$anonfun$apply$3.apply(BasicDirectives.scala:32)
    at akka.http.scaladsl.server.directives.FutureDirectives$$anonfun$onComplete$1$$anonfun$apply$1$$anonfun$apply$2.apply(FutureDirectives.scala:26)
    at akka.http.scaladsl.server.directives.FutureDirectives$$anonfun$onComplete$1$$anonfun$apply$1$$anonfun$apply$2.apply(FutureDirectives.scala:26)
    at akka.http.scaladsl.util.FastFuture$$anonfun$transformWith$extension0$1.apply(FastFuture.scala:37)
    at akka.http.scaladsl.util.FastFuture$$anonfun$transformWith$extension0$1.apply(FastFuture.scala:37)
    at akka.http.scaladsl.util.FastFuture$.akka$http$scaladsl$util$FastFuture$$strictTransform$1(FastFuture.scala:41)
    at akka.http.scaladsl.util.FastFuture$$anonfun$transformWith$extension1$1.apply(FastFuture.scala:51)
    at akka.http.scaladsl.util.FastFuture$$anonfun$transformWith$extension1$1.apply(FastFuture.scala:50)
    at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
    at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
    at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:90)
    at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:88)
    at akka.dispatch.BatchingExecutor$BlockableBatch$$anonfun$run$1.apply(BatchingExecutor.scala:88)
    at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
    at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:88)
    at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:401)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    ... 2 more
```

The top line corresponds to [this require](https://github.com/akka/akka/blob/b07d5da74bdcb016b6f2deaea6bce2047d9709bd/akka-actor/src/main/scala/akka/dispatch/BatchingExecutor.scala#L105).

I will attempt to isolate the issue, but I don't really understand what the "require" is checking, so I may have trouble in picking out the relevant code.

This started happening after upgrading to 1.0-RC1; very similar code ran fine in 1.0-M4.

I notice that this code differs between:
 * The [version in 2.3.10](https://github.com/akka/akka/blob/b07d5da74bdcb016b6f2deaea6bce2047d9709bd/akka-actor/src/main/scala/akka/dispatch/BatchingExecutor.scala#L105) that HTTP 1.0-RC1 relies on via SBT / Maven depedencies
 * The [version in the 1.0-RC1 tag](https://github.com/akka/akka/blob/releasing-akka-stream-and-http-experimental-1.0-RC1/akka-actor/src/main/scala/akka/dispatch/BatchingExecutor.scala#L100) which isn't actually included in the RC1 release 

This seems extremely suspicious to me. Has there been some mistake in the release process, or perhaps in my SBT imports?

My SBT looks like:
```
  val akkaV = "2.3.7"
  val akkaStreamsV = "1.0-RC1-myFork-0.1"
  val scalaV = "2.11.6"

  val deps = Seq(
    "ch.qos.logback"      %  "logback-classic"                % "1.1.2",
    "com.google.guava"    %  "guava"                          % "17.0",
    "com.typesafe.akka"   %% "akka-slf4j"                     % akkaV,
    "com.typesafe.akka"   %  "akka-http-core-experimental_2.11" % akkaStreamsV,
    "com.typesafe.akka"   %  "akka-http-spray-json-experimental_2.11" % akkaStreamsV,
    "com.typesafe.akka"   %% "akka-http-testkit-scala-experimental" % akkaStreamsV % "test",
    "com.typesafe.akka"   %% "akka-testkit"                   % akkaV,
   ....
```

My IVY dependency report looks like:

module | revision
---- | ----
akka-http-spray-json-experimental_2.11 by com.typesafe.akka | 1.0-RC1-myFork-0.1 
--- akka-http-scala-experimental_2.11 by com.typesafe.akka | 1.0-RC1-myFork-0.1 
------ akka-http-core-experimental_2.11 by com.typesafe.akka | 1.0-RC1-myFork-0.1
--------- akka-stream-experimental_2.11 by com.typesafe.akka | 1.0-RC1-myFork-0.1
------------ reactive-streams by org.reactivestreams | 1.0.0.RC5 
------------ akka-actor_2.11 by com.typesafe.akka | 2.3.10 

(note that "myFork-0.1" is a branch build including #17332. I didn't change anything else, but it's possible that I messed that up somehow)

**Comment by jrudolph** [comment]
Thanks, @RichardBradley for the report.

> This seems extremely suspicious to me.

I don't think that's suspicious. The akka-actor code in the release-2.3-dev branch isn't really used for anything (and hasn't been for a long time). It's just an artifact of the (custom) branch and module setup of the akka project at this particular point in time.

For now, I don't even think this is an akka-http issue. Whatever the `require` is supposed to do, it doesn't provide a proper error message or any explanation how to deal with the issue if it's a user error so it's at least a documentation problem of akka-actor.

**Comment by jrudolph** [comment]
> This seems extremely suspicious to me.

You may still be somewhat on the right track: going from M4 to RC1 the dependency on akka-actor was updated from 2.3.9 to 2.3.10 which contained this commit f74a670c83c573cdccc9f49629d9217e2c18bff3 which is probably related and also contains the diff you looked at above. (The diff between release-2.3 and release-2.3-dev contains all changes in akka-actor of the last year, that's why I was skeptical before.)

Can you try forcing akka-actor to 2.3.9 and see if it fixes the issue for you?

**Comment by drewhk** [comment]
Yep, so release-2.3-dev should not be used to build anything else than streams and http. It is supposed to be merged cleanly against release-2.3. Any fixes in ordinary Akka go into release-2.3, and not into the dev branch. The streams and http projects depend on released akka artifacts.

**Comment by codingismy11to7** [comment]
I want to +1 this issue, we're not using akka-http at all, but using some blocking on an `ask` in a short-lived install-time utility; I've got coworkers that are seeing the same 4 top lines of the stack trace (haven't seen it myself, but they are seeing it fairly consistently; going to rollback to 2.3.9).

**Comment by rkuhn** [comment]
The fix is in #17390.

**Comment by rkuhn** [comment]
The fix has been merged to release-2.3 and will be released as part of version 2.3.11 shortly. We can close this ticket as soon as release-2.3-dev depends on 2.3.11.

**Comment by RichardBradley** [comment]
> Can you try forcing akka-actor to 2.3.9 and see if it fixes the issue for you?

(This has fixed things for me, thanks.)

**Comment by rkuhn** [comment]
This has been fixed in 2.3.11, if you still see it with that version then by all means reopen this issue—it is highly recommended to update since that version fixed another regression that was introduced in 2.3.7.
